### DISCUSSION TITLE
The UNDO Command

### DISCUSSION THREAD

**Discussion by tajmone** [alan-if/alan-i18n#47] [Tech Stuff] [answered]
I noticed that the Spanish library contains in the [`meta_sesión.i`][meta_sesión.i] module (formerly `meta.i`) a commented-out definition of the UNDO command, left over from the original pALANte library:

```alan
-- Syntax undo = undo.
--
-- Meta verb undo
--   Does "Unfortunately you cannot 'undo' commands in this game."
-- End verb.
```

Which seems to indicate that originally there was no UNDO feature in ALAN, and that it was introduced somewhere along the timeline in which pALANte was being translated (possibly when ALAN 3 came out).

@thoni56, do you recall any details about the history of the UNDO command?

Also, it looks like this is the only player command which is hard-wired into the ALAN interpreter, since it works even if the adventure doesn't provide any `SYNTAX` and `VERB` definitions for it (unlike `LOOK`, `QUIT`, etc.).

The only reference I've found to this in _The ALAN Manual_ is in [§5.6. _Undoing_]:

> The player command to restore a previous game state is handled directly by the interpreter. It must consist of the single word undo.

I was wondering whether we should add Spanish `SYNONYMS` for UNDO, or whether this command is so universal in IF that it doesn't deserve a translation. @Rich15, what do you think?

But I'm also curious about this hard-wired command, and whether it's even possible to provide a custom `VERB` for it (e.g. to customize its behaviour). Since there's no `UNDO` keyword to invoke the operation (like we do for `LOOK`, `SAVE`, etc.) I'm not sure how a VERB would handle this.

And what happens if a game implements a custom `UNDO` command, then? Does it take over the hard-wired UNDO, shadowing it and preventing its use? or does it cause a compile error?

I'm just curious about this, since it falls under the same category of the `hero` hard-wired special actor (which _can_ be customized).

But I also think that it's something worth mentioning in the library documentation, as well as adding a few comments about it in the library module that deals with META VERBs (as a reminder to authors and maintainers).

[meta_sesión.i]: https://github.com/alan-if/alan-i18n/blob/main/alan_es/Foundation/meta_sesión.i "View source file"

[§5.6. _Undoing_]: https://alan-if.github.io/alan-docs/manual-beta/manual.html#_undoing "ALAN Manual » §5.6. Undoing"

**Comment by thoni56** [comment, marked as answer]
> Which seems to indicate that originally there was no UNDO feature in ALAN, and that it was introduced somewhere along the timeline in which pALANte was being translated (possibly when ALAN 3 came out).
> 
> @thoni56, do you recall any details about the history of the UNDO command?
> 

That is a correct assumption. I'm not sure about the timelines, but that would explain the comment.

> Also, it looks like this is the only player command which is hard-wired into the ALAN interpreter, since it works even if the adventure doesn't provide any `SYNTAX` and `VERB` definitions for it (unlike `LOOK`, `QUIT`, etc.).

Also true.
 
> The only reference I've found to this in _The ALAN Manual_ is in [§5.6. _Undoing_](https://alan-if.github.io/alan-docs/manual-beta/manual.html#_undoing):
> 
> > The player command to restore a previous game state is handled directly by the interpreter. It must consist of the single word undo.
> 

This probably should be explanded upon in the manual.

> 
> But I'm also curious about this hard-wired command, and whether it's even possible to provide a custom `VERB` for it (e.g. to customize its behaviour). Since there's no `UNDO` keyword to invoke the operation (like we do for `LOOK`, `SAVE`, etc.) I'm not sure how a VERB would handle this.

The UNDO must be handled by the interpreter before anything else happens or the undo-ing will not work given the current design. (Everything is possible with software, but with any reasonable effort and complication to the code.)

> And what happens if a game implements a custom `UNDO` command, then? Does it take over the hard-wired UNDO, shadowing it and preventing its use? or does it cause a compile error?

A game implementing an `undo` verb will compile without any error messages (probably should give at least a warning). As the UNDO is hard-wired in the interpreter the "real" undo will still be available when running the game. (Just tested it.)

> 
> I'm just curious about this, since it falls under the same category of the `hero` hard-wired special actor (which _can_ be customized).
> 
> But I also think that it's something worth mentioning in the library documentation, as well as adding a few comments about it in the library module that deals with META VERBs (as a reminder to authors and maintainers).
